lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/driver/ResourceManager.java

```diff
@@ -199,41 +203,87 @@ void onResourceLaunchRequest(
+  }
+
+  /**
+  /**
    * Checks the allocation queue for new allocations and if there are any
    * satisfies them.
    */
   private void checkRequestQueue() {
 
-    if (this.theContainers.hasContainerAvailable() && this.requestQueue.hasOutStandingRequests()) {
-
-      // Record the satisfaction of one request and get its details.
-      final ResourceRequestEvent requestEvent = this.requestQueue.satisfyOne();
-
-      // Allocate a Container
-      final Container container = this.theContainers.allocateOne(
-              requestEvent.getMemorySize().orElse(this.defaultMemorySize),
-              requestEvent.getVirtualCores().orElse(this.defaultNumberOfCores));
-
-      // Tell the receivers about it
-      final ResourceAllocationEvent alloc =
-          ResourceAllocationEventImpl.newBuilder()
-              .setIdentifier(container.getContainerID())
-              .setNodeId(container.getNodeID())
-              .setResourceMemory(container.getMemory())
-              .setVirtualCores(container.getNumberOfCores())
-              .setRackName(container.getRackName())
-              .build();
-
-      LOG.log(Level.FINEST, "Allocating container: {0}", container);
-      this.allocationHandler.onNext(alloc);
-
+    if (requestQueue.hasOutStandingRequests()) {
+      final ResourceRequest resourceRequest = requestQueue.head();
+      final ResourceRequestEvent requestEvent = resourceRequest.getRequestProto();
+      final List<String> rackNames = getRackNamesOrDefault(requestEvent.getRackNameList());
+      validateRackNames(rackNames);
+      boolean allocated = false;
+      for (final String rackName : rackNames) {
+        if (theContainers.hasContainerAvailable(rackName)) {
+          requestQueue.satisfyOne();
+          // Allocate a Container
+          // Not taking into account the node names for now
+          final Container container = this.theContainers.allocateOne(
+                  requestEvent.getMemorySize().orElse(this.defaultMemorySize),
+                  requestEvent.getVirtualCores().orElse(this.defaultNumberOfCores),
+                  rackName);
+
+          // Tell the receivers about it
+          final ResourceAllocationEvent alloc =
+              ResourceAllocationEventImpl.newBuilder()
+                  .setIdentifier(container.getContainerID())
+                  .setNodeId(container.getNodeID())
+                  .setResourceMemory(container.getMemory())
+                  .setVirtualCores(container.getNumberOfCores())
+                  .setRackName(container.getRackName())
+                  .build();
+
+          LOG.log(Level.FINEST, "Allocating container: {0}", container);
+          this.allocationHandler.onNext(alloc);
+          allocated = true;
+        }
+        // if we allocated the container, we break and update the status.
+        if (allocated) {
+          LOG.log(Level.FINEST, "Allocated on rack {0}", rackName);
+          break;
+        } else {
```

Comment:
I'm unclear on the semantics of the list of rack names. This would suggest that if we can't allocate on the first rack given, we give up. Is that the desired behavior? I understood the list of racknames initially as a priority list: We'll match the first possible one on the list. This would also allow us to treat relaxed locality by appending the `ANY` rack to the end of the list.